csharp/src/Google.Protobuf/ParsingPrimitives.cs

```diff
@@ -638,17 +636,68 @@ public static string ReadRawString(ref ReadOnlySpan<byte> buffer, ref ParserInte
             }
 #endif
 
-            var decoder = CodedOutputStream.Utf8Encoding.GetDecoder();
+            return ReadStringSlow(ref buffer, ref state, length);
+        }
+
+        /// <summary>
+        /// Reads a string assuming that it is spread across multiple spans in a <see cref="ReadOnlySequence{T}"/>.
+        /// </summary>
+        private static string ReadStringSlow(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, int length)
+        {
+#if GOOGLE_PROTOBUF_SUPPORT_FAST_STRING
+            if (IsDataAvailable(ref state, length))
+            {
+                if (state.stringDecoder == null)
+                {
+                    state.stringDecoder = CodedOutputStream.Utf8Encoding.GetDecoder();
+                }
+
+                // We need to decode bytes incrementally across multiple spans.
+                int maxCharLength = CodedOutputStream.Utf8Encoding.GetMaxCharCount(length);
+                char[] charArray = ArrayPool<char>.Shared.Rent(maxCharLength);
```

Comment:
I'm not convinced that using the array pool here is always going to what the user wants. Is there a precedent of using array pools for such a basic functionality (like parsing a string in what is a general purpose library).

My worries is that using the array pool here can mess with e.g. user app's concurrency model or their own uses of the arrayPool (we don't know what the user app does and user does not have an option to opt-out). I'm fine with this if there is sufficient precedent that this is ok to do.

Reply:
Yes, this pattern is commonly used in .NET Core runtime when you temporarily need an array.

https://github.com/dotnet/runtime/blob/365c91674a32b6b54181c8248c32cf38a85069b2/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L14-L36

Using a stackalloc array for small strings like they do here is a good improvement.